File: Backend/api/models/loans.py
from django.db import models
from .spaces import Space, SpaceMember
from .contacts import Contact
from decimal import Decimal

class LoanDirection(models.TextChoices):
    GIVEN = 'GIVEN', 'Given'
    TAKEN = 'TAKEN', 'Taken'

class LoanStatus(models.TextChoices):
    DRAFT = 'DRAFT', 'Draft'
    ACTIVE = 'ACTIVE', 'Active'
    CLOSED = 'CLOSED', 'Closed'

class ClosureReason(models.TextChoices):
    FULLY_PAID = 'FULLY_PAID', 'Fully Paid'
    SETTLED = 'SETTLED', 'Settled'
    WRITTEN_OFF = 'WRITTEN_OFF', 'Written Off'
    MANUALLY_CLOSED = 'MANUALLY_CLOSED', 'Manually Closed'

class InterestType(models.TextChoices):
    NONE = 'NONE', 'No Interest'
    FIXED = 'FIXED', 'Fixed Amount'
    FLAT = 'FLAT', 'Flat Rate'
    REDUCING_BALANCE = 'REDUCING_BALANCE', 'Reducing Balance'
    COMPOUND = 'COMPOUND', 'Compound'
    CUSTOM = 'CUSTOM', 'Custom Interest'

class FixedInterestFrequency(models.TextChoices):
    ONE_TIME = 'ONE_TIME', 'One Time'
    RECURRING = 'RECURRING', 'Recurring'

class RatePeriod(models.TextChoices):
    DAY = 'DAY', 'Daily'
    WEEK = 'WEEK', 'Weekly'
    MONTH = 'MONTH', 'Monthly'
    YEAR = 'YEAR', 'Yearly'

class InterestTiming(models.TextChoices):
    COLLECTED_UPFRONT = 'COLLECTED_UPFRONT', 'Collected Upfront'
    DEDUCTED_FROM_DISBURSEMENT = 'DEDUCTED_FROM_DISBURSEMENT', 'Deducted From Disbursement'
    PAYABLE_AT_END = 'PAYABLE_AT_END', 'Payable at End'
    PAYABLE_PERIODICALLY = 'PAYABLE_PERIODICALLY', 'Payable Periodically'

class InterestRateBehavior(models.TextChoices):
    FIXED = 'FIXED', 'Fixed'
    VARIABLE = 'VARIABLE', 'Variable'
    PROMOTIONAL = 'PROMOTIONAL', 'Promotional'

class RepaymentType(models.TextChoices):
    ONE_TIME = 'ONE_TIME', 'One-time (Bullet)'
    EMI = 'EMI', 'EMI'
    INTEREST_ONLY = 'INTEREST_ONLY', 'Interest Only'
    PRINCIPAL_ONLY = 'PRINCIPAL_ONLY', 'Principal Only'
    FLEXIBLE = 'FLEXIBLE', 'Flexible'
    CUSTOM_INSTALLMENTS = 'CUSTOM_INSTALLMENTS', 'Custom Installments'

class PaymentFrequency(models.TextChoices):
    DAILY = 'DAILY', 'Daily'
    WEEKLY = 'WEEKLY', 'Weekly'
    BI_WEEKLY = 'BI_WEEKLY', 'Bi-weekly'
    MONTHLY = 'MONTHLY', 'Monthly'
    QUARTERLY = 'QUARTERLY', 'Quarterly'

class PaymentTimingRule(models.TextChoices):
    SCHEDULED = 'SCHEDULED', 'Scheduled'
    ANYTIME = 'ANYTIME', 'Anytime'

class AdvancePaymentMode(models.TextChoices):
    CARRY_FORWARD_CREDIT = 'CARRY_FORWARD_CREDIT', 'Carry Forward Credit'
    RECALCULATE_SCHEDULE = 'RECALCULATE_SCHEDULE', 'Recalculate Schedule'

class PenaltyType(models.TextChoices):
    NONE = 'NONE', 'None'
    FIXED = 'FIXED', 'Fixed Penalty'
    PERCENTAGE = 'PERCENTAGE', 'Percentage Penalty'
    DAILY_LATE_FEE = 'DAILY_LATE_FEE', 'Daily Late Fee'
    MONTHLY_LATE_FEE = 'MONTHLY_LATE_FEE', 'Monthly Late Fee'
    EXTRA_INTEREST = 'EXTRA_INTEREST', 'Extra Interest'

class DisbursementLabel(models.TextChoices):
    ORIGINAL = 'ORIGINAL', 'Original'
    TOP_UP = 'TOP_UP', 'Top-up'
    ADDITIONAL_BORROWING = 'ADDITIONAL_BORROWING', 'Additional Borrowing'

class RateHistoryTrigger(models.TextChoices):
    INITIAL = 'INITIAL', 'Initial'
    VARIABLE_CHANGE = 'VARIABLE_CHANGE', 'Variable Change'
    PROMO_EXPIRY = 'PROMO_EXPIRY', 'Promotional Expiry'
    RESTRUCTURING = 'RESTRUCTURING', 'Restructuring'

class WaiverType(models.TextChoices):
    INTEREST = 'INTEREST', 'Interest'
    PENALTY = 'PENALTY', 'Penalty'

class ScheduleLineStatus(models.TextChoices):
    PENDING = 'PENDING', 'Pending'
    PAID = 'PAID', 'Paid'


class Loan(models.Model):
    space = models.ForeignKey(Space, on_delete=models.CASCADE, related_name='loans')
    contact = models.ForeignKey(Contact, on_delete=models.RESTRICT, related_name='loans')
    direction = models.CharField(max_length=10, choices=LoanDirection.choices)
    status = models.CharField(
        max_length=10,
        choices=LoanStatus.choices,
        default=LoanStatus.DRAFT
    )
    closure_reason = models.CharField(
        max_length=20,
        choices=ClosureReason.choices,
        null=True,
        blank=True
    )
    closure_note = models.TextField(null=True, blank=True)
    closed_at = models.DateTimeField(null=True, blank=True)
    closed_by = models.ForeignKey(
        SpaceMember,
        null=True,
        blank=True,
        on_delete=models.PROTECT,
        related_name='loans_closed'
    )
    written_off_amount = models.DecimalField(max_digits=14, decimal_places=2, null=True, blank=True)
    reopened_at = models.DateTimeField(null=True, blank=True)
    created_by = models.ForeignKey(SpaceMember, on_delete=models.PROTECT, related_name='loans_created')
    created_at = models.DateTimeField(auto_now_add=True)
    updated_at = models.DateTimeField(auto_now=True)

    principal_amount = models.DecimalField(max_digits=14, decimal_places=2)
    start_date = models.DateTimeField()
    first_due_date = models.DateField(null=True, blank=True)
    tenure_periods = models.PositiveSmallIntegerField(null=True, blank=True)

    interest_type = models.CharField(max_length=20, choices=InterestType.choices)
    rate_value = models.DecimalField(max_digits=10, decimal_places=4, null=True, blank=True)
    rate_period = models.CharField(max_length=10, choices=RatePeriod.choices, null=True, blank=True)
    fixed_interest_amount = models.DecimalField(max_digits=14, decimal_places=2, null=True, blank=True)
    fixed_interest_frequency = models.CharField(
        max_length=20,
        choices=FixedInterestFrequency.choices,
        null=True,
        blank=True
    )
    interest_timing = models.CharField(
        max_length=30,
        choices=InterestTiming.choices,
        default=InterestTiming.PAYABLE_PERIODICALLY
    )
    net_disbursed_amount = models.DecimalField(max_digits=14, decimal_places=2, null=True, blank=True)
    interest_rate_behavior = models.CharField(
        max_length=20,
        choices=InterestRateBehavior.choices,
        default=InterestRateBehavior.FIXED
    )
    promo_rate = models.DecimalField(max_digits=10, decimal_places=4, null=True, blank=True)
    promo_period_days = models.PositiveSmallIntegerField(null=True, blank=True)

    repayment_type = models.CharField(max_length=20, choices=RepaymentType.choices)
    has_balloon_final_payment = models.BooleanField(default=False)
    payment_frequency = models.CharField(
        max_length=20,
        choices=PaymentFrequency.choices,
        null=True,
        blank=True
    )
    payment_timing_rule = models.CharField(
        max_length=20,
        choices=PaymentTimingRule.choices,
        default=PaymentTimingRule.SCHEDULED
    )

    advance_payment_mode = models.CharField(max_length=30, choices=AdvancePaymentMode.choices)
    advance_credit_balance = models.DecimalField(max_digits=14, decimal_places=2, default=Decimal('0.00'))
    penalty_type = models.CharField(
        max_length=20,
        choices=PenaltyType.choices,
        default=PenaltyType.NONE
    )
    penalty_value = models.DecimalField(max_digits=10, decimal_places=4, null=True, blank=True)
    grace_period_days = models.PositiveSmallIntegerField(default=0)

    class Meta:
        db_table = 'loans'
        app_label = 'api'
        indexes = [
            models.Index(fields=['space']),
            models.Index(fields=['space', 'status']),
            models.Index(fields=['space', 'contact']),
            models.Index(fields=['space', 'direction', 'status']),
        ]

    def __str__(self):
        return f"Loan #{self.id} ({self.contact.name} - {self.direction})"


class Disbursement(models.Model):
    space = models.ForeignKey(Space, on_delete=models.CASCADE, related_name='+')
    loan = models.ForeignKey(Loan, on_delete=models.CASCADE, related_name='disbursements')
    amount = models.DecimalField(max_digits=14, decimal_places=2)
    disbursement_date = models.DateTimeField()
    sequence_no = models.PositiveSmallIntegerField()
    label = models.CharField(
        max_length=30,
        choices=DisbursementLabel.choices,
        default=DisbursementLabel.ORIGINAL
    )
    created_by = models.ForeignKey(SpaceMember, on_delete=models.PROTECT, related_name='+')
    created_at = models.DateTimeField(auto_now_add=True)

    class Meta:
        db_table = 'disbursements'
        app_label = 'api'


class LoanRateHistory(models.Model):
    space = models.ForeignKey(Space, on_delete=models.CASCADE, related_name='+')
    loan = models.ForeignKey(Loan, on_delete=models.CASCADE, related_name='rate_history')
    effective_from = models.DateField()
    rate_value = models.DecimalField(max_digits=10, decimal_places=4)
    rate_period = models.CharField(max_length=10, choices=RatePeriod.choices)
    trigger = models.CharField(max_length=30, choices=RateHistoryTrigger.choices)
    reason = models.TextField(null=True, blank=True)
    created_by = models.ForeignKey(SpaceMember, on_delete=models.PROTECT, related_name='+')
    created_at = models.DateTimeField(auto_now_add=True)

    class Meta:
        db_table = 'loan_rate_history'
        app_label = 'api'


class LoanTenureExtension(models.Model):
    space = models.ForeignKey(Space, on_delete=models.CASCADE, related_name='+')
    loan = models.ForeignKey(Loan, on_delete=models.CASCADE, related_name='tenure_extensions')
    added_periods = models.PositiveSmallIntegerField()
    tenure_periods_before = models.PositiveSmallIntegerField()
    tenure_periods_after = models.PositiveSmallIntegerField()
    reason = models.TextField()
    performed_by = models.ForeignKey(SpaceMember, on_delete=models.PROTECT, related_name='+')
    created_at = models.DateTimeField(auto_now_add=True)

    class Meta:
        db_table = 'loan_tenure_extensions'
        app_label = 'api'


class LoanMoratorium(models.Model):
    space = models.ForeignKey(Space, on_delete=models.CASCADE, related_name='+')
    loan = models.ForeignKey(Loan, on_delete=models.CASCADE, related_name='moratoriums')
    pause_start_date = models.DateField()
    pause_end_date = models.DateField()
    interest_free = models.BooleanField(default=False)
    reason = models.TextField()
    performed_by = models.ForeignKey(SpaceMember, on_delete=models.PROTECT, related_name='+')
    created_at = models.DateTimeField(auto_now_add=True)

    class Meta:
        db_table = 'loan_moratoriums'
        app_label = 'api'


class LoanWaiver(models.Model):
    space = models.ForeignKey(Space, on_delete=models.CASCADE, related_name='+')
    loan = models.ForeignKey(Loan, on_delete=models.CASCADE, related_name='waivers')
    waiver_type = models.CharField(max_length=10, choices=WaiverType.choices)
    waived_amount = models.DecimalField(max_digits=14, decimal_places=2)
    reason = models.TextField()
    performed_by = models.ForeignKey(SpaceMember, on_delete=models.PROTECT, related_name='+')
    created_at = models.DateTimeField(auto_now_add=True)

    class Meta:
        db_table = 'loan_waivers'
        app_label = 'api'


class RepaymentScheduleLine(models.Model):
    space = models.ForeignKey(Space, on_delete=models.CASCADE, related_name='+')
    loan = models.ForeignKey(Loan, on_delete=models.CASCADE, related_name='schedule_lines')
    schedule_version = models.PositiveSmallIntegerField(default=1)
    line_no = models.PositiveSmallIntegerField()
    due_date = models.DateField()
    principal_due = models.DecimalField(max_digits=14, decimal_places=2)
    interest_due = models.DecimalField(max_digits=14, decimal_places=2, default=Decimal('0.00'))
    status = models.CharField(
        max_length=10,
        choices=ScheduleLineStatus.choices,
        default=ScheduleLineStatus.PENDING
    )
    is_current_version = models.BooleanField(default=True)
    superseded_by_type = models.CharField(max_length=30, null=True, blank=True)
    superseded_by_id = models.BigIntegerField(null=True, blank=True)
    is_custom_line = models.BooleanField(default=False)
    created_at = models.DateTimeField(auto_now_add=True)

    class Meta:
        db_table = 'repayment_schedule_lines'
        app_label = 'api'
        indexes = [
            models.Index(fields=['loan', 'is_current_version', 'status', 'due_date']),
        ]

File: Frontend/src/api/loans.ts
import { apiClient } from '@/lib/apiClient';

export interface Loan {
  id: number;
  contact_id: number;
  contact_name?: string; // Embedded by serializer (edge case #5)
  direction: 'GIVEN' | 'TAKEN';
  principal_amount: string;
  start_date: string;
  first_due_date: string | null;
  tenure_periods: number | null;

  interest_type: 'NONE' | 'FIXED' | 'FLAT' | 'REDUCING_BALANCE' | 'COMPOUND' | 'CUSTOM';
  rate_value: string | null;
  rate_period: 'DAY' | 'WEEK' | 'MONTH' | 'YEAR' | null;
  fixed_interest_amount: string | null;
  fixed_interest_frequency: string | null;
  interest_timing: 'UPFRONT' | 'DEDUCTED_FROM_DISBURSEMENT' | 'PAYABLE_PERIODICALLY' | 'AT_END';
  net_disbursed_amount: string | null;
  interest_rate_behavior: 'FIXED' | 'VARIABLE' | 'PROMOTIONAL';
  promo_rate: string | null;
  promo_period_days: number | null;

  repayment_type: 'ONE_TIME' | 'EMI' | 'INTEREST_ONLY' | 'PRINCIPAL_ONLY' | 'FLEXIBLE' | 'CUSTOM_INSTALLMENTS';
  payment_frequency: string | null;
  payment_timing_rule: 'SCHEDULED' | 'ANYTIME';

  advance_payment_mode: 'CARRY_FORWARD_CREDIT' | 'RECALCULATE_SCHEDULE' | null;
  penalty_type: 'NONE' | 'FIXED' | 'PERCENTAGE' | 'EXTRA_INTEREST';
  penalty_value: string | null;
  grace_period_days: number | null;

  status: 'DRAFT' | 'ACTIVE' | 'CLOSED';
  outstanding_balance: string;
  is_overdue: boolean;
  accrued_penalty_to_date: string;
  advance_credit_balance: string;
  written_off_amount?: string;
  closure_reason?: 'FULLY_PAID' | 'SETTLED' | 'WRITTEN_OFF' | 'MANUALLY_CLOSED' | null;
  closure_date?: string | null;
  created_at: string;
}

export interface Disbursement {
  id: number;
  sequence_no: number;
  amount: string;
  disbursement_date: string;
  label: 'ORIGINAL' | 'TOP_UP' | 'ADDITIONAL_BORROWING';
  created_at: string;
}

export interface ScheduleLine {
  id: number;
  due_date: string;
  principal_due: string;
  interest_due: string;
  principal_paid: string;
  interest_paid: string;
  penalty_paid: string;
  status: 'PAID' | 'PARTIALLY_PAID' | 'PENDING';
  schedule_version: number;
  is_current_version: boolean;
  is_overdue: boolean; // Computed client-side
}

export interface RestructureEvent {
  id: number;
  event_type: 'RATE_CHANGE' | 'TENURE_EXTENSION' | 'MORATORIUM' | 'WAIVER';
  description: string;
  actor_name: string;
  timestamp: string;
  reason: string;
  details: any;
}

export interface LoanListParams {
  direction?: 'GIVEN' | 'TAKEN';
  status?: 'DRAFT' | 'ACTIVE' | 'CLOSED';
  is_overdue?: boolean;
  closure_reason?: 'WRITTEN_OFF';
  page?: number;
}

// Loans List & CRUD
export const listLoans = async (
  spaceId: number,
  params?: LoanListParams
): Promise<{ count: number; results: Loan[] }> => {
  const response = await apiClient.get(`/spaces/${spaceId}/loans/`, { params });
  return response.data;
};

export const createLoan = async (spaceId: number, data: Partial<Loan>): Promise<Loan> => {
  const response = await apiClient.post(`/spaces/${spaceId}/loans/`, data);
  return response.data;
};

export const getLoan = async (spaceId: number, loanId: number): Promise<Loan> => {
  const response = await apiClient.get(`/spaces/${spaceId}/loans/${loanId}/`);
  return response.data;
};

export const updateLoan = async (spaceId: number, loanId: number, data: Partial<Loan>): Promise<Loan> => {
  const response = await apiClient.patch(`/spaces/${spaceId}/loans/${loanId}/`, data);
  return response.data;
};

// Lifecycle Actions
export const activateLoan = async (
  spaceId: number,
  loanId: number
): Promise<{ loan: Loan; warnings?: string[] }> => {
  const response = await apiClient.post(`/spaces/${spaceId}/loans/${loanId}/activate/`);
  return response.data;
};

export const closeLoan = async (
  spaceId: number,
  loanId: number,
  closureReason: 'FULLY_PAID' | 'MANUALLY_CLOSED',
  closureNote?: string
): Promise<Loan> => {
  const response = await apiClient.post(`/spaces/${spaceId}/loans/${loanId}/close/`, {
    closure_reason: closureReason,
    closure_note: closureNote,
  });
  return response.data;
};

export const closeEarlyLoan = async (
  spaceId: number,
  loanId: number,
  closureDate?: string
): Promise<Loan> => {
  const response = await apiClient.post(`/spaces/${spaceId}/loans/${loanId}/close-early/`, {
    closure_date: closureDate,
  });
  return response.data;
};

export const reopenLoan = async (spaceId: number, loanId: number, reason: string): Promise<Loan> => {
  const response = await apiClient.post(`/spaces/${spaceId}/loans/${loanId}/reopen/`, { reason });
  return response.data;
};

export const changeAdvanceMode = async (
  spaceId: number,
  loanId: number,
  mode: 'CARRY_FORWARD_CREDIT' | 'RECALCULATE_SCHEDULE'
): Promise<{
  advance_payment_mode: 'CARRY_FORWARD_CREDIT' | 'RECALCULATE_SCHEDULE';
  credit_applied: string;
  schedule_version: number;
  status: string;
}> => {
  const response = await apiClient.post(`/spaces/${spaceId}/loans/${loanId}/change-advance-mode/`, {
    advance_payment_mode: mode,
  });
  return response.data;
};

export const addFieldNote = async (spaceId: number, loanId: number, note: string): Promise<void> => {
  await apiClient.post(`/spaces/${spaceId}/loans/${loanId}/notes/`, { note });
};

// Disbursements
export const listDisbursements = async (spaceId: number, loanId: number): Promise<Disbursement[]> => {
  const response = await apiClient.get(`/spaces/${spaceId}/loans/${loanId}/disbursements/`);
  return response.data;
};

export const recordDisbursement = async (
  spaceId: number,
  loanId: number,
  data: { amount: string; disbursement_date: string; label: 'TOP_UP' | 'ADDITIONAL_BORROWING' }
): Promise<Disbursement> => {
  const response = await apiClient.post(`/spaces/${spaceId}/loans/${loanId}/disbursements/`, data);
  return response.data;
};

// Schedule
export const getSchedule = async (
  spaceId: number,
  loanId: number,
  includeSuperseded: boolean = false
): Promise<ScheduleLine[]> => {
  const response = await apiClient.get(`/spaces/${spaceId}/loans/${loanId}/schedule/`, {
    params: { include_superseded: includeSuperseded },
  });
  return response.data;
};

export const setCustomScheduleLines = async (
  spaceId: number,
  loanId: number,
  lines: Array<{ due_date: string; principal_due: number; interest_due: number }>
): Promise<{ warnings?: string[] }> => {
  const response = await apiClient.post(
    `/spaces/${spaceId}/loans/${loanId}/schedule/custom-lines/`,
    lines
  );
  return response.data;
};

// Restructuring Action APIs
export const restructureRateChange = async (
  spaceId: number,
  loanId: number,
  data: { effective_from: string; rate_value: number; rate_period: string; reason: string }
): Promise<void> => {
  await apiClient.post(`/spaces/${spaceId}/loans/${loanId}/restructure/rate-change/`, data);
};

export const restructureExtendTenure = async (
  spaceId: number,
  loanId: number,
  data: { added_periods: number; reason: string }
): Promise<void> => {
  await apiClient.post(`/spaces/${spaceId}/loans/${loanId}/restructure/extend-tenure/`, data);
};

export const restructureMoratorium = async (
  spaceId: number,
  loanId: number,
  data: { pause_start_date: string; pause_end_date: string; interest_free: boolean; reason: string }
): Promise<void> => {
  await apiClient.post(`/spaces/${spaceId}/loans/${loanId}/restructure/moratorium/`, data);
};

export const restructureWaiveInterest = async (
  spaceId: number,
  loanId: number,
  data: { waived_amount: number; reason: string }
): Promise<void> => {
  await apiClient.post(`/spaces/${spaceId}/loans/${loanId}/restructure/waive-interest/`, data);
};

export const restructureWaivePenalty = async (
  spaceId: number,
  loanId: number,
  data: { waived_amount: number; reason: string }
): Promise<void> => {
  await apiClient.post(`/spaces/${spaceId}/loans/${loanId}/restructure/waive-penalty/`, data);
};

export const getRestructuringHistory = async (spaceId: number, loanId: number): Promise<RestructureEvent[]> => {
  const response = await apiClient.get(`/spaces/${spaceId}/loans/${loanId}/restructure/history/`);
  return response.data;
};
